File: test/domain/recovery/save_buffer_to_recovery_test.dart
// TDD — TASK-05 (M2) + TASK-06 (M5): SaveBufferToRecovery use case.
//
// Test strategy:
//   A fake RecoveryRepository records the CALL ORDER across save() and trim()
//   so TASK-06 tests can assert trim(10) always runs AFTER save — and never
//   when save throws or when the trim-empty guard fires.
//
//   M2 scenarios (§5.1.2, EC-M2-02):
//   1. Non-empty text  → delegates RAW text to repo.save; returns the sentinel.
//   2. Empty string    → returns null; repo.save NEVER called.
//   3. Whitespace-only → returns null; repo.save NEVER called.
//   4. Padded text     → trim is non-empty, so delegates the RAW (un-trimmed)
//                        text to repo.save; NOT the trimmed version.
//
//   M5 scenarios (FR-M5-03, §5.1.4):
//   5. Non-empty text  → call order is ['save:<text>', 'trim:10'].
//   6. save() throws   → exception propagates; trim NOT called.
//   7. Whitespace-only → returns null; neither save nor trim called.
//
// Format: given_<context>_when_<action>_then_<outcome>

import 'dart:io';

import 'package:buffer/domain/recovery/recovery_note.dart';
import 'package:buffer/domain/recovery/recovery_repository.dart';
import 'package:buffer/domain/recovery/save_buffer_to_recovery.dart';
import 'package:flutter_test/flutter_test.dart';

// ---------------------------------------------------------------------------
// Fake
// ---------------------------------------------------------------------------

class _FakeRecoveryRepository implements RecoveryRepository {
  /// Ordered record of every call made. Entries have the form:
  ///   `save:<text>`  — for save(text)
  ///   `trim:<keep>`  — for trim(keep)
  final List<String> callLog = [];

  /// When non-null, save() throws this exception instead of returning.
  Object? saveError;

  final File sentinel;

  _FakeRecoveryRepository(this.sentinel);

  @override
  Future<File> save(String text) async {
    if (saveError != null) throw saveError!;
    callLog.add('save:$text');
    return sentinel;
  }

  @override
  Future<void> trim(int keep) async {
    callLog.add('trim:$keep');
  }

  // --- stubs for M5 members not exercised by this use case ---

  @override
  Future<List<RecoveryNote>> list() async => const [];

  @override
  Future<String> read(RecoveryNote note) async => '';

  @override
  Future<void> delete(RecoveryNote note) async {}

  @override
  Future<void> deleteAll() async {}

  // Defect-B sync stub — not exercised by use-case unit tests.
  @override
  File saveSync(String text, {int keep = 10}) => sentinel;
}

// ---------------------------------------------------------------------------
// Tests
// ---------------------------------------------------------------------------

void main() {
  late File sentinel;
  late _FakeRecoveryRepository fakeRepo;
  late SaveBufferToRecovery useCase;

  setUp(() {
    sentinel = File('sentinel.txt');
    fakeRepo = _FakeRecoveryRepository(sentinel);
    useCase = SaveBufferToRecovery(fakeRepo);
  });

  group('SaveBufferToRecovery', () {
    // -------------------------------------------------------------------------
    // M2: Non-empty text → delegates raw text, returns sentinel (FR-M2-13, §5.1.2)
    // -------------------------------------------------------------------------

    test(
      'given_non_empty_text_when_called_then_delegates_raw_text_to_repository_and_returns_file',
      () async {
        final result = await useCase('hello');

        expect(fakeRepo.callLog, contains('save:hello'));
        expect(result, equals(sentinel));
      },
    );

    // -------------------------------------------------------------------------
    // M2: Empty string → null, zero I/O (EC-M2-02)
    // -------------------------------------------------------------------------

    test(
      'given_empty_string_when_called_then_returns_null_with_zero_repository_calls',
      () async {
        final result = await useCase('');

        expect(result, isNull);
        expect(fakeRepo.callLog, isEmpty);
      },
    );

    // -------------------------------------------------------------------------
    // M2: Whitespace-only → null, zero I/O (EC-M2-02)
    // -------------------------------------------------------------------------

    test(
      'given_whitespace_only_text_when_called_then_returns_null_with_zero_repository_calls',
      () async {
        final result = await useCase('   \n\t');

        expect(result, isNull);
        expect(fakeRepo.callLog, isEmpty);
      },
    );

    // -------------------------------------------------------------------------
    // M2: Padded text → trim decides empty/non-empty, but RAW text is delegated
    // -------------------------------------------------------------------------

    test(
      'given_padded_non_empty_text_when_called_then_delegates_raw_untrimed_text_to_repository',
      () async {
        final result = await useCase('  hi  ');

        expect(fakeRepo.callLog, contains('save:  hi  '));
        expect(result, equals(sentinel));
      },
    );

    // -------------------------------------------------------------------------
    // M5: trim(10) called AFTER save, in that order (FR-M5-03, §5.1.4)
    // -------------------------------------------------------------------------

    test(
      'given_non_empty_text_when_called_then_trim10_invoked_strictly_after_save',
      () async {
        await useCase('hello');

        expect(fakeRepo.callLog, equals(['save:hello', 'trim:10']));
      },
    );

    // -------------------------------------------------------------------------
    // M5: save throws → trim NOT called; exception propagates (§5.1.4)
    // -------------------------------------------------------------------------

    test(
      'given_repository_save_throws_when_called_then_exception_propagates_and_trim_is_not_called',
      () async {
        fakeRepo.saveError = const FileSystemException('disk full');

        await expectLater(
          () => useCase('hello'),
          throwsA(isA<FileSystemException>()),
        );
        expect(fakeRepo.callLog, isEmpty);
      },
    );

    // -------------------------------------------------------------------------
    // M5: whitespace-only → M2 guard fires; neither save nor trim invoked
    // -------------------------------------------------------------------------

    test(
      'given_whitespace_only_text_when_called_then_neither_save_nor_trim_invoked',
      () async {
        final result = await useCase('   ');

        expect(result, isNull);
        expect(fakeRepo.callLog, isEmpty);
      },
    );
  });
}

File: shared/src/commonTest/kotlin/com/paolosantucci/foglietto/shared/recovery/SaveBufferToRecoveryTest.kt
package com.paolosantucci.foglietto.shared.recovery

import kotlin.test.Test
import kotlin.test.assertEquals
import kotlin.test.assertFailsWith
import kotlin.test.assertNull
import kotlin.test.assertTrue

// ---------------------------------------------------------------------------
// Hand-written fake — records the exact call sequence for assertion.
//
// `internal` visibility: required so the test class (public by default in
// kotlin.test) can access the fake without a private-in-file restriction on
// the JVM test runner.
// ---------------------------------------------------------------------------

/**
 * Hand-written fake [RecoveryRepository] that records each call as a
 * structured string entry in [callLog].
 *
 * Entry format:
 *   `"save:<text>"`  — produced by [save]
 *   `"trim:<keep>"`  — produced by [trim]
 *
 * When [saveError] is non-null, [save] throws it instead of recording.
 *
 * The remaining members ([list], [read], [delete], [deleteAll]) are stubs
 * that satisfy the interface but are never exercised by these use-case tests.
 */
internal class FakeRecoveryRepositoryForUseCase : RecoveryRepository {

    val callLog: MutableList<String> = mutableListOf()

    /** When non-null, [save] throws this instead of recording. */
    var saveError: Throwable? = null

    /** The path string returned by [save] when no error is set. */
    val sentinelPath: String = "sentinel.txt"

    override fun save(text: String): String {
        saveError?.let { throw it }
        callLog.add("save:$text")
        return sentinelPath
    }

    override fun trim(keep: Int) {
        callLog.add("trim:$keep")
    }

    // --- stubs not exercised by SaveBufferToRecovery tests ---

    override fun list(): List<RecoveryNote> = emptyList()

    override fun read(path: String): String? = null

    override fun delete(path: String) {}

    override fun deleteAll() {}
}

// ---------------------------------------------------------------------------
// SaveBufferToRecoveryTest — 6 cases from the Dart oracle.
// (Drops callSync / _trimSync helper tests: collapsed into one synchronous save.)
// ---------------------------------------------------------------------------

class SaveBufferToRecoveryTest {

    // -----------------------------------------------------------------------
    // Case 1 — Empty string → null, ZERO repository calls (EC-06, FR-07)
    // -----------------------------------------------------------------------

    @Test
    fun given_empty_string_when_invoked_then_returns_null_with_zero_repository_calls() {
        val fakeRepo = FakeRecoveryRepositoryForUseCase()
        val useCase = SaveBufferToRecovery(fakeRepo)

        val result = useCase("")

        assertNull(result, "invoke(\"\") must return null")
        assertTrue(
            fakeRepo.callLog.isEmpty(),
            "Expected zero repository calls, got: ${fakeRepo.callLog}"
        )
    }

    // -----------------------------------------------------------------------
    // Case 2 — Whitespace-only → null, ZERO repository calls (EC-06, FR-07)
    // -----------------------------------------------------------------------

    @Test
    fun given_whitespace_only_text_when_invoked_then_returns_null_with_zero_repository_calls() {
        val fakeRepo = FakeRecoveryRepositoryForUseCase()
        val useCase = SaveBufferToRecovery(fakeRepo)

        val result = useCase("   \n\t ")

        assertNull(result, "invoke(whitespace-only) must return null")
        assertTrue(
            fakeRepo.callLog.isEmpty(),
            "Expected zero repository calls, got: ${fakeRepo.callLog}"
        )
    }

    // -----------------------------------------------------------------------
    // Case 3 — Raw un-trimmed text delegated to save (FR-08)
    // The trim-guard uses text.trim() ONLY to decide empty/non-empty;
    // the RAW (un-trimmed) string is what reaches repository.save.
    // -----------------------------------------------------------------------

    @Test
    fun given_padded_non_empty_text_when_invoked_then_delegates_raw_untrimed_text_to_repository() {
        val fakeRepo = FakeRecoveryRepositoryForUseCase()
        val useCase = SaveBufferToRecovery(fakeRepo)

        useCase(" hi ")

        assertTrue(
            fakeRepo.callLog.contains("save: hi "),
            "repository.save must receive the raw un-trimmed text ' hi ', got: ${fakeRepo.callLog}"
        )
    }

    // -----------------------------------------------------------------------
    // Case 4 — After successful save, trim is called with exactly 10 (FR-08)
    // -----------------------------------------------------------------------

    @Test
    fun given_non_empty_text_when_invoked_then_trim_called_with_10() {
        val fakeRepo = FakeRecoveryRepositoryForUseCase()
        val useCase = SaveBufferToRecovery(fakeRepo)

        useCase("hello")

        assertTrue(
            fakeRepo.callLog.contains("trim:10"),
            "repository.trim must be called with keep=10, got: ${fakeRepo.callLog}"
        )
    }

    // -----------------------------------------------------------------------
    // Case 5 — save called BEFORE trim (strict order, FR-08)
    // -----------------------------------------------------------------------

    @Test
    fun given_non_empty_text_when_invoked_then_save_is_called_strictly_before_trim() {
        val fakeRepo = FakeRecoveryRepositoryForUseCase()
        val useCase = SaveBufferToRecovery(fakeRepo)

        useCase("hello")

        assertEquals(
            listOf("save:hello", "trim:10"),
            fakeRepo.callLog,
            "Call order must be [save:hello, trim:10]"
        )
    }

    // -----------------------------------------------------------------------
    // Case 6 — save throws → exception propagates, trim NOT called (FR-09, EC-05)
    // -----------------------------------------------------------------------

    @Test
    fun given_repository_save_throws_when_invoked_then_exception_propagates_and_trim_not_called() {
        val fakeRepo = FakeRecoveryRepositoryForUseCase()
        val useCase = SaveBufferToRecovery(fakeRepo)
        val expectedError = RuntimeException("disk full")
        fakeRepo.saveError = expectedError

        val thrown = assertFailsWith<RuntimeException>(
            message = "save's exception must propagate unchanged"
        ) {
            useCase("hello")
        }
        assertEquals(expectedError, thrown)
        assertTrue(
            fakeRepo.callLog.isEmpty(),
            "trim must NOT be called when save throws, got: ${fakeRepo.callLog}"
        )
    }
}
